test_rclcpp/test/test_subscription.cpp

```diff
@@ -129,53 +156,46 @@ TEST(CLASSNAME(test_subscription, RMW_IMPLEMENTATION), subscription_and_spinning
     ASSERT_EQ(1, counter);
 
     // while four messages have been published one callback should be triggered here
-    printf("spin_node_once(nonblocking) - callback (2) expected\n");
-    executor.spin_node_once(node, std::chrono::milliseconds(0));
-    if (counter == 1) {
-      // give the executor thread time to process the event
-      std::this_thread::sleep_for(std::chrono::milliseconds(25));
-      printf("spin_node_once(nonblocking) - callback (2) expected - trying again\n");
-      executor.spin_node_once(node, std::chrono::milliseconds(0));
-    }
+    printf("spin_until_future_complete(short timeout) - callback (2) expected\n");
+    sub_called = std::promise<void>();
+    sub_called_future = sub_called.get_future();
+    wait_for_future(executor, sub_called_future, 10_ms);
     ASSERT_EQ(2, counter);
 
     // check for next pending call
-    printf("spin_node_once(nonblocking) - callback (3) expected\n");
-    executor.spin_node_once(node, std::chrono::milliseconds(0));
-    if (counter == 2) {
-      // give the executor thread time to process the event
-      std::this_thread::sleep_for(std::chrono::milliseconds(25));
-      printf("spin_node_once(nonblocking) - callback (3) expected - trying again\n");
-      executor.spin_node_once(node, std::chrono::milliseconds(0));
-    }
+    printf("spin_until_future_complete(short timeout) - callback (3) expected\n");
+    sub_called = std::promise<void>();
+    sub_called_future = sub_called.get_future();
+    wait_for_future(executor, sub_called_future, 10_ms);
     ASSERT_EQ(3, counter);
 
-    // check for all remaning calls
-    printf("spin_node_some() - callbacks (4 and 5) expected\n");
-    executor.spin_node_some(node);
-    if (counter == 3 || counter == 4) {
-      // give the executor thread time to process the event
-      std::this_thread::sleep_for(std::chrono::milliseconds(25));
-      printf("spin_node_some() - callback (%s) expected - trying again\n",
-        counter == 3 ? "4 and 5" : "5");
-      executor.spin_node_once(node, std::chrono::milliseconds(0));
-    }
+    // check for next pending call
+    printf("spin_until_future_complete(short timeout) - callback (4) expected\n");
+    sub_called = std::promise<void>();
+    sub_called_future = sub_called.get_future();
+    wait_for_future(executor, sub_called_future, 10_ms);
+    ASSERT_EQ(4, counter);
+
+    // check for last pending call (blocking)
+    printf("spin_until_future_complete() - callback (5) expected\n");
+    sub_called = std::promise<void>();
+    sub_called_future = sub_called.get_future();
+    wait_for_future(executor, sub_called_future, fail_after_timeout);
     ASSERT_EQ(5, counter);
   }
   // the subscriber goes out of scope and should be not receive any callbacks anymore
 
-  // wait a moment for everything to initialize
-  // TODO(gerkey): fix nondeterministic startup behavior
-  rclcpp::utilities::sleep_for(1_ms);
-
```

Comment:
I removed this because even if the underlying middleware is still connected and the `rmw` subscription gets notified as ready for the "6" we publish below, it should never make it to the `rclcpp` subscription callback since that class has been destroyed.